File: [프로그래머스] 주식가격.py
def solution(prices):
    #몇초인지
    n = len(prices)
    ans = [0]*n
    for i in range(n):
        for j in range(i+1,n):#만약 0초면 1초부터 비교
            if prices[i] > prices[j]:#주식가격이 떨어지면
                ans[i] += j-i
                break
        else:
            ans[i] = n-i-1
    return ans




--------------------------------------------------

# 2022- 05 -05
# 이게 더 빠름

def solution(prices):
    stack = []
    answer = [ 0 for i in range(len(prices))]
    
    for time, price in enumerate(prices):
        
        if len(stack) > 0:    
            while stack:
                if stack[-1][1] > price:
                    (s_time, s_price) = stack.pop()
                    answer[s_time] = time - s_time
                else: #주식 안떨어졌으면 
                    break
        
        stack.append((time,price))

    
    #끝까지 안떨어진 값들 정리
    cur_time = len(prices) - 1
    while stack:
        (time,price) = stack.pop()
        answer[time] = cur_time - time
        
    return answer

File: [프로그래머스] 주식가격.java
import java.util.*;


class Solution {
    public int[] solution(int[] prices) {
        
        //주식가격이 떨어지지 않았다 에서 -> stack 을 생각하는 키워드!
        
        int[] ans = new int[prices.length];
        Stack<Integer> stack = new Stack();
        
        for (int i = 0; i < prices.length; i++){
            while (!stack.isEmpty() && prices[i] < prices[stack.peek()]) {
                ans[stack.peek()] = i - stack.peek(); //주식가격이 떨어짐
                stack.pop(); //답을 구했기 때문에 stack에서 제거
            }
            stack.push(i);
        }
        
        while(!stack.isEmpty()){//값을 구하지 못한 index == 끝까지 가격이 떨어지지 않은 주식
            ans[stack.peek()] = (prices.length-1) - stack.peek();
            stack.pop();
        }
        
        return ans;
    }

}
